### PR TITLE
[ZEN-4770] Suggestion to use fixed versions for transitive dependencies in GenTarget POMs

### PR DESCRIPTION
**Pull request by ghillairet**
This PR demonstrates how the project can be reorganized to avoid bringing all dependencies at once when using a specific gentemplate.

It removes from the module genflow-api the dependency to rapid-ml to a new module called genflow-api-rapidml. This latter module is used only by gentemplates module that need rapidml. Other gentemplate modules for swagger and openapi thus do not pull anymore dependencies on EMF/Xtext and Eclipse, thus improving their build time.

### CODE REVIEW

**Comment by ghillairet** [comment]
See PR in RAPID-ML to remove version ranges for EMF https://github.com/RepreZen/RAPID-ML/pull/23

**Comment by tedepstein** [comment]
Suggested testing strategy:
* @ghillairet  will create an API Studio PR that uses this.
* Create a new new project, remove unneeded GenFlow dependency (RAPID-ML or OpenAPI) from project POM.
* Open the POM in Eclipse and switch to the dependency tree view. Inspect the tree to ensure that there are no ranges shown, and everything resolves to the expected version.

**Comment by tedepstein** [comment]
Separate follow-up issues (to be prioritized):
* Create "POM packages" that are source-specific: one each for OpenAPI v2, v3, and RAPID-ML.  Each POM package would just be a wrapper around other dependencies, specifically those GenTemplate providers that have GenTemplates available for that source. 
* Change default project POM generator so instead of creating one dependency on all of GenFlow, it creates several more granular dependencies, either by source type (as above) or directly by provider. Make it easier for users who want to remove unneeded GenGemplate dependencies for faster, simpler builds. 
* To facilitate the above, add logic to GenTarget creation and/or execution to make sure that the project POM includes all dependencies needed for the selected GenTarget(s). Offer to add the missing dependencies to the project POM if needed.

File: pom.xml [hunk 1]
```diff
@@ -4,483 +4,493 @@
...
+				<groupId>org.yaml</groupId>
+				<artifactId>snakeyaml</artifactId>
+				<version>1.23</version>
+			</dependency>
+			<dependency>
+				<groupId>org.eclipse.xtend</groupId>
+				<artifactId>org.eclipse.xtend.lib</artifactId>
+				<version>2.9.2</version>
+			</dependency>
+			<dependency>
```
Comment (tedepstein) [inline pom.xml]: Why are Eclipse dependencies added here? Were they formerly included as transitive dependencies of RAPID-ML, and are still needed by other parts of GenFlow?

**Comment by ghillairet** [comment]
We still have several dependencies to eclipse plugins in the project `genflow-common` (see https://github.com/RepreZen/GenFlow/blob/task/ZEN-4770/modules/gentemplates/common/pom.xml#L31) 

I am not sure what is the need for it. I can see an import of `IFile` in the class https://github.com/RepreZen/GenFlow/blob/task/ZEN-4770/modules/gentemplates/common/src/main/java/com/reprezen/genflow/common/observe/GeneratedArtifact.java I can't find any use of this class in GenFlow or API Studio (except in some tests). And also a use of eclipse `FileLocator` here https://github.com/RepreZen/GenFlow/blob/task/ZEN-4770/modules/gentemplates/common/src/main/java/com/reprezen/genflow/common/codegen/GenModulesInfo.java#L75 

That seems the only need for those dependencies. I will look if those are still needed or if it is only there for legacy reasons.

**Comment by ghillairet** [comment]
Also `swagger-nsswag` should be split into 2 projects as it contains gentemplates for both `swagger` and `rapidml`.

**Comment by tedepstein** [comment]
@ghillairet , 

> Also swagger-nsswag should be split into 2 projects as it contains gentemplates for both swagger and rapidml.

We should just drop the RAPID-ML GenTemplates from the NSwag module. The RAPID-ML GenTemplates just wrap around a 2-phase transformation:
* RAPID-ML --> OpenAPI v2 via the RAPID-XChange Contract GenTemplate
* OpenAPI v2 --> .NET or TypeScript via NSwag. 

We did this for a specific customer who were using RAPID-ML and wanted to generate for .NET. They are no longer using NSwag, so they won't miss these GenTemplates. 

Scenarios like this are better addressed through the general multi-step chaining mechanism that should be documented as part of the GenFlow architecture and feature set.

**Comment by ghillairet** [comment]
Ready to review.

File: modules/standard-gentemplates/pom.xml [hunk 2]
```diff
@@ -52,71 +52,18 @@
     <dependencies>
         <dependency>
             <groupId>com.reprezen.genflow</groupId>
-            <artifactId>openapi3-doc</artifactId>
+            <artifactId>reprezen-openapi-gentemplates</artifactId>
```
Comment (tedepstein) [inline modules/standard-gentemplates/pom.xml]: We shouldn't need this directly anymore. Already in `openapi-gentemplates`

**Comment by tedepstein** [comment]
Testing strategy is to go down the hierarchy for each category of gentemplates. Hierarchy looks like this:

* standard-gentemplates.
    * openapi-gentemplates
        * openapi-generator-gentemplates
            * (individual gentemplates)
        * swagger-codegen-gentemplates
            * (individual gentemplates)
        * reprezen-openapi-gentemplates
            * (individual gentemplates)
    * reprezen-rapidml-gentemplates
        * (individual gentemplates)

* For each GenTemplate Provider Group (e.g. `reprezen-openapi-gentemplates`)
    * Create a GenTarget using one of the GenTemplates in this group (e.g. RepreZen HTML Documentation), run the GenTemplate and make sure it works.
    * Modify the project POM file to use the aggregate group, e.g. "openapi-generators"

**Comment by nbhusare** [comment]
CNF while generating Java code for Pet Store example. The dependencies in the pom are not changed. They are kept AIS while generation

```
[main] WARN com.reprezen.genflow.api.template.GenTemplateRegistry - Could not retrieve gentemplate; skipping
java.lang.NoClassDefFoundError: org/eclipse/xtext/xbase/lib/CollectionLiterals
	at com.reprezen.genflow.rapidml.jaxrs.JavaXtendTemplate.<clinit>(JavaXtendTemplate.java:24)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at com.reprezen.genflow.api.template.builders.BuilderUtil.getClass(BuilderUtil.java:17)
	at com.reprezen.genflow.api.template.builders.OutputItemBuilder.using(OutputItemBuilder.java:33)
	at com.reprezen.genflow.api.template.builders.OutputItemBuilder.using(OutputItemBuilder.java:29)
	at com.reprezen.genflow.rapidml.jaxrs.JaxRsGenTemplate.configure(JaxRsGenTemplate.java:30)
	at com.reprezen.genflow.api.template.GenTemplate.init(GenTemplate.java:192)
	at com.reprezen.genflow.api.template.GenTemplate.getAlsoKnownAsIds(GenTemplate.java:135)
	at com.reprezen.genflow.api.template.GenTemplateInfo.extractIdInfo(GenTemplateInfo.java:157)
	at com.reprezen.genflow.api.template.GenTemplateInfo.<init>(GenTemplateInfo.java:45)
	at com.reprezen.genflow.api.template.GenTemplateRegistry.register(GenTemplateRegistry.java:49)
	at com.reprezen.genflow.api.template.GenTemplateRegistry.scan(GenTemplateRegistry.java:102)
	at com.reprezen.genflow.api.template.GenTemplateRegistry.getGenTemplate(GenTemplateRegistry.java:84)
	at com.reprezen.genflow.api.template.GenTemplateRegistry.getDefaultGenTemplate(GenTemplateRegistry.java:92)
	at com.reprezen.genflow.api.target.GenTargetUtils.getProvider(GenTargetUtils.java:142)
	at com.reprezen.genflow.api.target.GenTargetUtils.execute(GenTargetUtils.java:89)
	at com.reprezen.genflow.api.target.GenTargetUtils.execute(GenTargetUtils.java:103)
	at com.reprezen.genflow.api.util.GeneratorLauncher.run(GeneratorLauncher.java:87)
	at com.reprezen.genflow.api.util.GeneratorLauncher.main(GeneratorLauncher.java:45)
	at com.modelsolv.reprezen.generators.api.util.GeneratorLauncher.main(GeneratorLauncher.java:5)
Caused by: java.lang.ClassNotFoundException: org.eclipse.xtext.xbase.lib.CollectionLiterals
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 21 more
```

**Comment by nbhusare** [comment]
Tested the following use-cases for OpenAPI V2 (Swagger) and V3. The code generation for RAPID-ML fails with the error as explained above https://github.com/RepreZen/GenFlow/pull/55#issuecomment-485581932

- genflow/standard-gentemplates 
  - Generate code for Java Client and RepreZen HTML
- genflow/openapi-codegen
  - Generate code for Java Client and RepreZen HTML
- genflow/reprezen-openapi-gentemplate
  - Generate code for RepreZen HTML, Swagger UI v2, v3

**Comment by tedepstein** [comment]
@neerajbhusare , 

> Tested the following use-cases for OpenAPI v2 (Swagger) and v3.

Did these run successfully? 

> * genflow/standard-gentemplates
>   * Generate code for Java, C# Client, and RepreZen HTML

Should work for all supported GenTemplates. Were the Java and C# Client tests run with OAG?

> * genflow/openapi-codegen
>   * Generate code for Java, C# Client, and RepreZen HTML

If the project POM dependency was ib `genflow/openapi-codegen`, I would expect the test to work _only_ with OAG GenTemplates. Did it still work with the RepreZen HTML GenTemplate?  Did you happen to try this from the command line, or only from the IDE?

> The version range of `openapi-parser` in the root pom has been removed. It is not fixed to version [4.0.3](https://github.com/RepreZen/GenFlow/pull/55/files#diff-600376dffeb79835ede4a0b285078036R274). 

Did you mean _not fixed_ or _now fixed_?

**Comment by nbhusare** [comment]
@tedepstein 

>Tested the following use-cases for OpenAPI v2 (Swagger) and v3.
>>Did these run successfully?

Yes, the above-mentioned tests ran successfully. While I was going through your comments, I tried one more test which failed. The details are explained in the following comments.

> - genflow/standard-gentemplates 
     - Generate code for Java, C# Client, and RepreZen HTML
>>Should work for all supported GenTemplates. Were the Java and C# Client tests run with OAG?

Yes, the tests were run with OAG.

> - genflow/openapi-codegen
>   - Generate code for Java, C# Client, and RepreZen HTML
>>If the project POM dependency was on genflow/openapi-codegen, I would expect the test to work 
    only with OAG GenTemplates. Did it still work with the RepreZen HTML GenTemplate? Did you  
    happen to try this from the command line, or only from the IDE?

I meant - `genflow/openapi-*gentemplates*`. I corrected the same in my comment above. 

When I replaced the above dependency with `genflow/openapi-generator` (for OpenAPI v3) and `genflow/swagger-codegen` (for OpenAPI v2), the code-gen fails with the following error. I am assuming this is a valid test.

```
Exception in thread "main" java.lang.RuntimeException: Failed to execute all GenTargets (some may have been executed)
	at com.reprezen.genflow.api.util.GeneratorLauncher.main(GeneratorLauncher.java:46)
	at com.modelsolv.reprezen.generators.api.util.GeneratorLauncher.main(GeneratorLauncher.java:5)
```

PS - I am testing using the API Studio.

> The version range of openapi-parser in the root pom has been removed. It is not fixed to version 4.0.3.
> >Did you mean not fixed or now fixed?

My bad. I meant *now* fixed.

**Comment by ghillairet** [comment]
@nbhusare Can you post the pom file where you get those errors https://github.com/RepreZen/GenFlow/pull/55#issuecomment-485581932 I can't reproduce them.

**Comment by nbhusare** [comment]
@ghillairet The first error can be reproduced with the pom that is generated as part of the API Studio project. For the second one, I am using the same pom, just replaced the `standard-gentemplates` dependency with `swagger-codegen`. 

```
<dependencies>
		<dependency>
			<groupId>com.reprezen.genflow</groupId>
			<artifactId>swagger-codegen</artifactId>
			<version>1.3.1-ZEN-4770-SNAPSHOT</version>
			<type>pom</type>
		</dependency>
		<dependency>
			<groupId>com.reprezen.rapidml</groupId>
			<artifactId>com.reprezen.rapidml</artifactId>
			<version>0.0.10-ZEN-4770-SNAPSHOT</version>
		</dependency>
		<dependency>
			<groupId>com.reprezen.rapidml</groupId>
			<artifactId>com.reprezen.rapidml.model</artifactId>
			<version>0.0.10-ZEN-4770-SNAPSHOT</version>
		</dependency>
		<dependency>
			<groupId>com.reprezen.jsonoverlay</groupId>
			<artifactId>jsonoverlay</artifactId>
			<version>4.0.3</version>
		</dependency>
		<dependency>
			<groupId>com.reprezen.kaizen</groupId>
			<artifactId>openapi-parser</artifactId>
			<version>4.0.3</version>
		</dependency>
		<dependency>
			<groupId>io.swagger</groupId>
			<artifactId>swagger-codegen</artifactId>
			<version>2.4.0</version>
		</dependency>
		<dependency>
			<groupId>org.openapitools</groupId>
			<artifactId>openapi-generator</artifactId>
			<version>3.3.4</version>
		</dependency>
	</dependencies>
```

**Comment by ghillairet** [comment]
@nbhusare I still can't reproduce. Did you at some point made a local build of GenFlow branch/ZEN-4770? If yes maybe delete it from your local repository or re-build with latest changes.

@tedepstein Can you reproduce those issues?

**Pull request merged**